src/components/games/gamequeue.cpp

```diff
@@ -0,0 +1,118 @@
+}
+
+void qb::QueueComponent::register_actions(Actions<>& actions)
+{
+    using namespace std::placeholders;
+    register_all(
+        actions,
+        std::make_pair("startqueue", (ActionCallback)std::bind(&qb::QueueComponent::add_queue, this, _1, _2, _3)),
+        std::make_pair("stopqueue", (ActionCallback)std::bind(&qb::QueueComponent::remove_queue, this, _1, _2, _3))
+    );
+}
+
+qb::Result qb::QueueComponent::add_queue(const std::string& cmd, const api::Message& msg, Bot& bot)
+{
+    const auto& channel = msg.channel;
+    std::vector<std::string> tokens = qb::parse::split(cmd);
+    if( tokens.size() != 4)
+    {
+        bot.send("Incorrect number of parameters given. Proper format : !qb queue <game> <queue_name> <max_size>", channel);
+        return qb::Result::ok();
+    }
+    else
+    {
+        const auto game = tokens[1];
+        const auto name = tokens[2];
+        const auto max_size = std::stoi(tokens[3]);
+
+        Queue new_queue = Queue(name, msg, game, max_size);
+        send_yn_message(new_queue, bot, "Queueing a game of " + game + " called " + name + " with 0 players. (Max " + tokens[3] + ")", channel);
+        return qb::Result::ok();
+    }
+}
+
+qb::Result qb::QueueComponent::remove_queue(const std::string& cmd, const api::Message& msg, Bot& bot)
+{
+    const auto& channel = msg.channel;
+    std::vector<std::string> tokens = qb::parse::split(cmd);
+    if(tokens.size() != 2)
+    {
+        bot.send("Incorrect number of parameters given. Proper format : !qb dequeue <name>", channel);
+        return qb::Result::ok();
+    }
+    else
+    {
+        const auto name = tokens[1];
+        active_queues.erase(msg.id);
+        bot.send("Queue called " + name + " removed.", channel);
+        return qb::Result::ok();
+    }
+}
+
+nlohmann::json qb::QueueComponent::send_yn_message(Queue& queue, Bot& bot, const std::string& message, const std::string& channel)
+{
+    qb::log::point("A queue message.");
+    const auto resp = send_removable_message(bot, message, channel);
+    if(resp.empty()) return {};
+    active_queues.emplace(resp["id"], queue);
+    bot.on_message_id(resp["id"], bind_action(&qb::QueueComponent::add_yn_reaction, this));
+    return resp;
+
+}
+qb::Result qb::QueueComponent::add_yn_reaction( const std::string& message_id, const api::Message& message, Bot& bot)
+{
+    if (const auto& yes_emote = qb::parse::emote_snowflake(qb::fileio::get_emote("yes")); yes_emote)
+    {
+        bot.get_context()->put(qb::endpoints::reaction(message.channel, message_id, *yes_emote), {});
+    }
+    if (const auto& no_emote = qb::parse::emote_snowflake(qb::fileio::get_emote("no")); no_emote)
+    {
+        bot.get_context()->put(qb::endpoints::reaction(message.channel, message_id, *no_emote), {});
+    }
+
+    bot.on_message_reaction(message, [this, em = qb::fileio::get_emote("yes"), message_id, message](
```

Comment:
wait I'm confused does this compile? you're passing `message` and not `message_id` to `on_message_reaction` let me go see...
wow it does actually work that way that's really cool who wrote that code it's so safe and fancy

Reply:
also the `em = qb::fileio::get_emote("yes")` is copied right, so this is weird and we should make a decision for this

we basically have two options and they're equally easy (actually the current implementation is harder than the other option, marginally)

1. we cache the emote so that even if the yes-emote is changed (via configure/restart/whatever) the emote that we reacted with is the emote to react yes _to that message_. pros: this is sorta fancy and also you never need to hunt for the emote, like if you change it after making a queue you don't have to manually add the first new reaction you can just continue using the old one. cons: not really expected behaviour, I guess? ehhhh okay the way we have it is probably correct. other con is it wastes memory
2. we don't cache the emote and alway scompare to `get_emote("yes")` technically slower due to more indirection but then it does consume a small amount more memory. now if you do `assign yes xyz` then you need to react to all old queues with `xyz` instead of what was already on the message. this doesn't actually make sense and is bad.
3. nevermind we don't have to make a decision the way it is right now is definitely correct, good stuff